File: python/src/trailproof/chain.py
"""Hash chain engine with canonical JSON serialization."""

import hashlib
import json
from dataclasses import asdict

from trailproof.types import TrailEvent

GENESIS_HASH = "0" * 64

_EXCLUDED_FIELDS = {"hash", "signature"}


def canonical_json(event: TrailEvent) -> str:
    """Serialize an event to canonical JSON for hashing.

    Produces deterministic output:
    - Keys sorted alphabetically (recursive for nested objects)
    - Compact format: no whitespace
    - Excludes "hash" and "signature" fields
    - Excludes None fields
    - UTF-8 encoding

    Args:
        event: The trail event to serialize.

    Returns:
        A canonical JSON string.
    """
    raw = asdict(event)
    cleaned = _strip_excluded(raw)
    return json.dumps(cleaned, sort_keys=True, separators=(",", ":"), ensure_ascii=False)


def compute_hash(prev_hash: str, event: TrailEvent) -> str:
    """Compute the SHA-256 hash for an event in the chain.

    hash = SHA-256(prev_hash + canonical_json(event))

    Args:
        prev_hash: Hash of the previous event, or GENESIS_HASH for the first.
        event: The trail event to hash.

    Returns:
        Hex-encoded SHA-256 hash string.
    """
    payload = prev_hash + canonical_json(event)
    return hashlib.sha256(payload.encode("utf-8")).hexdigest()


def _strip_excluded(data: dict[str, object]) -> dict[str, object]:
    """Remove excluded fields and None values, recursing into nested dicts.

    Args:
        data: Dictionary to clean.

    Returns:
        A new dictionary with excluded fields and None values removed,
        nested dicts sorted recursively.
    """
    result: dict[str, object] = {}
    for key, value in data.items():
        if key in _EXCLUDED_FIELDS:
            continue
        if value is None:
            continue
        if isinstance(value, dict):
            result[key] = _strip_excluded(value)
        else:
            result[key] = value
    return result

File: typescript/src/chain.ts
/**
 * Hash chain engine with canonical JSON serialization.
 */

import { createHash } from "node:crypto";
import type { TrailEvent } from "./types.js";

/** Genesis hash: 64 zero characters. */
export const GENESIS_HASH = "0".repeat(64);

const EXCLUDED_FIELDS = new Set(["hash", "signature"]);

/**
 * Serialize an event to canonical JSON for hashing.
 *
 * Produces deterministic output:
 * - Keys sorted alphabetically (recursive for nested objects)
 * - Compact format: no whitespace
 * - Excludes "hash" and "signature" fields
 * - Excludes null/undefined fields
 * - UTF-8 encoding
 */
export function canonicalJson(event: TrailEvent): string {
  const raw: Record<string, unknown> = { ...event };
  const cleaned = stripExcluded(raw);
  return stableStringify(cleaned);
}

/**
 * Compute the SHA-256 hash for an event in the chain.
 *
 * hash = SHA-256(prev_hash + canonicalJson(event))
 */
export function computeHash(prevHash: string, event: TrailEvent): string {
  const payload = prevHash + canonicalJson(event);
  return createHash("sha256").update(payload, "utf-8").digest("hex");
}

/** Remove excluded fields and null/undefined values, recursing into nested objects. */
function stripExcluded(data: Record<string, unknown>): Record<string, unknown> {
  const result: Record<string, unknown> = {};
  for (const key of Object.keys(data)) {
    if (EXCLUDED_FIELDS.has(key)) continue;
    const value = data[key];
    if (value === null || value === undefined) continue;
    if (typeof value === "object" && !Array.isArray(value)) {
      result[key] = stripExcluded(value as Record<string, unknown>);
    } else {
      result[key] = value;
    }
  }
  return result;
}

/** JSON.stringify with sorted keys (recursive), compact format. */
function stableStringify(obj: unknown): string {
  if (obj === null || obj === undefined) return "null";
  if (typeof obj === "string") return JSON.stringify(obj);
  if (typeof obj === "number" || typeof obj === "boolean") return String(obj);
  if (Array.isArray(obj)) {
    return "[" + obj.map((item) => stableStringify(item)).join(",") + "]";
  }
  if (typeof obj === "object") {
    const record = obj as Record<string, unknown>;
    const keys = Object.keys(record).sort();
    const pairs = keys.map((k) => JSON.stringify(k) + ":" + stableStringify(record[k]));
    return "{" + pairs.join(",") + "}";
  }
  return String(obj);
}
